DOCSP-13320: Define a Realm Object Schema (.NET) (#669)
mongodb/docs-realm: source/dotnet/objects.txt

schema for their object type and validates objects whenever they're
created, modified, or deleted.

.. _dotnet-field-types:

Field Types
~~~~~~~~~~~

> {+client-database+} supports the following field data types:

Reviewer: It should also mention that the nullable versions of these types are supported.

.. literalinclude:: /examples/Schemas/Lists.cs
  :language: csharp

.. _dotnet-field-embedded:

Embedded Objects
~~~~~~~~~~~~~~~~

{+client-database+} provides the ability to nest objects within other
objects. This has several advantages:

- If using {+sync+}, objects will translate into MongoDB documents that
  follow a :manual:`denormalized data model </core/data-model-design/>`.

- When you delete an object that contains another object, the delete
  operation removes both objects from the {+realm+}, so unused objects
  don't accumulate in your {+realm+} file, taking up valuable space on
  user's mobile devices.

To embed an object, extend 
:dotnet-sdk:`EmbeddedObject <reference/Realms.EmbeddedObject.html>`
for the class that you'd like to nest within another class:

.. literalinclude:: /examples/Schemas/Embedded.cs
  :language: csharp

Then, any time you reference that class from another class,
{+client-database+} will embed the referenced class within the enclosing
class.

.. seealso::

   :ref:`Embedded Objects <dotnet-embedded-objects>`

.. _dotnet-annotations:

Annotations
------------

.. _dotnet-primary-key:

Primary Key
~~~~~~~~~~~

> A **primary key** is a field that uniquely identifies an object. You can

Reviewer: This list is missing `char` as well as the nullable versions of ObjectId and the numeric types.

create a primary key with any of the following types:

> - ``String``

Reviewer: This should be lowercase

- ``ObjectId``
- ``Integer`` or ``int``
- ``Long`` or ``long``

> - ``Short`` or ``short``

Reviewer: The capitalized versions of these types should be removed.

> - ``Byte`` or ``byte[]``

Reviewer: `byte[]` cannot be a primary key. `byte` can, but it has to be lowercase.

schema for their object type and validates objects whenever they're
created, modified, or deleted.

.. _dotnet-field-types:

Field Types
~~~~~~~~~~~

{+client-database+} supports the following field data types:

- ``Boolean`` or ``boolean``
- ``Integer`` or ``int``
- ``Short`` or ``short``
- ``Long`` or ``long``
- ``Byte`` or ``byte[]``
- ``Double`` or ``double``
- ``Float`` or ``float``
- ``String``
- ``Date``
- ``Decimal128`` from ``org.bson.types``
- ``ObjectId`` from ``org.bson.types``
- A ``RealmObject`` subclass (see :ref:`Relationship Annotations <dotnet-field-relationships>`)
- An ``EmbeddedObject`` class (see :ref:`Embedded Object Annotations <dotnet-field-relationships-embedded>`)

> - ``IList<T>`` where T is any of the previous data types

Reviewer 2: ```suggestion
- ``IList<T>`` where T is any of the previous data types (see :ref:`Lists <dotnet-field-lists>`)
```

The ``Byte``, ``Short``, ``Integer``, and ``Long`` types and their
lowercase primitive alternatives are all stored as ``Long`` values
within {+client-database+}. Similarly, {+client-database+} stores objects
of the ``Float`` and ``float`` types as type ``Double``.

``final`` and ``volatile`` fields are not supported, though you can
include fields with those modifiers if you :ref:`ignore
<dotnet-ignore-field>` those fields. If you choose to provide custom
constructors, you must declare a public constructor with no arguments.

.. _dotnet-field-lists:

Lists
~~~~~

Realm objects can contain lists of non-{+service-short+}-object data
types. You can model these collections with the standard .NET interface
`IList<T> <https://docs.microsoft.com/en-us/dotnet/api/system.collections.generic.ilist-1?view=net-5.0>`__,
where ``T`` can be the following types:

- ``String``
- ``Integer``
- ``Boolean``
- ``Float``
- ``Double``
- ``Short``
- ``Long``
- ``Byte``
- ``byte[]``

> - ``Date``

Reviewer 2: nits:
* These are in a different order from the previous section which makes it tougher to compare them
* Do lists also support `Decimal128` and `ObjectId`? AFAIK they do, i.e. they should support lists of any primitive type.

Maybe it'd be easier to just remove the list and say you can have lists of any of the primitive data types? i.e. you can have a list of anything except for RealmObject (which uses Relationships instead) or another List.

Reviewer: I think it would be confusing to say you can have a list of anything except RealmObject considering a relationship is defined precisely as a list of RealmObject 😅

Reviewer 2: Ah fair enough! That simplifies it to "a List of anything except for another List"

         :language: csharp

.. _dotnet-required-field:

Required Fields
~~~~~~~~~~~~~~~

You can make a field required to disallow ``null`` values in a field.
You can make any of the following types required:

- ``String``
- ``ObjectId``
- ``Integer``
- ``Long``
- ``Short``
- ``Byte`` or ``byte[]``
- ``Boolean``
- ``Float``
- ``Double``
- ``Date``

   .. literalinclude:: /examples/Schemas/Required.cs
         :language: csharp

.. warning:: .NET Types and Nullability

   In .NET, value types are non-nullable by default unless you explicitly
   add a ``?`` suffix to the type, and cannot have the ``[Required]`` annotation. 
   Reference types like ``String`` and ``Date`` 
   are nullable by default, and can have the annotation without 
   failing compilation. Fields with the ``RealmObject`` type are always
   nullable and cannot be required.

Optional Fields
~~~~~~~~~~~~~~~

In .NET, fields are nullable only if marked with with the 
C# `? operator <https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/builtin-types/nullable-value-types>`__ except
for the following types:

      - ``String``
      - ``Date``
      - ``ObjectId``

>       - ``Decimal128``

Reviewer 2: nit: let's totally de-indent this list

You can require any type that ends with the C# ``?``
operator, such as ``Int?``.

Any value type in C# can be made nullable with the ``?``, although by
default they are non-nullable without it. 
Nullable fields are optional by default in {+client-database+}, unless
otherwise specified with the :ref:`[Required] <dotnet-required-field>`
annotation. 

.. _dotnet-default-field-values:

Default Field Values
~~~~~~~~~~~~~~~~~~~~

To assign a default value to a field, use the built-in language features
to assign default values. In .NET, you can assign a default value in
the field declaration. 

   .. literalinclude:: /examples/Schemas/DefaultValues.cs

>          :language: csharp

Reviewer 2: Let's de-indent the literalinclude (and other similar ones on this page)

objects of the same type in a {+realm+}. 

.. note::
   
   Primary keys can be of type ``char``, ``short``, ``int``, ``long``, ``string``,

>    or ``MongoDB.Bson.ObjectId``.

Reviewer 2: This is redundant with the other list at the start of the section + doesn't really fit with the rest of the note. Can we remove it?

   Once you assign a property as a primary key, you cannot change it after adding an
   object of that type to a {+realm+}.
   
   The following example demonstrates how to designate a primary key in an object schema:
   
   .. literalinclude:: /examples/Schemas/PrimaryKey.cs
      :language: csharp

.. _dotnet-required-field:

Required Fields
~~~~~~~~~~~~~~~

You can make a field required to disallow ``null`` values in a field.
You can make any of the following types required:

- ``String``
- ``ObjectId``
- ``Integer``
- ``Long``
- ``Short``
- ``Byte`` or ``byte[]``
- ``Boolean``
- ``Float``
- ``Double``

> - ``Date``

Reviewer 2: Maybe it'd be easier to say what you **can't** make required? Then we avoid redundancy with the list of primitive data types

.. example::
   
   .. literalinclude:: /examples/Schemas/Required.cs
      :language: csharp

.. warning:: .NET Types and Nullability

   In .NET, value types are non-nullable by default unless you explicitly
   add a ``?`` suffix to the type, and cannot have the ``[Required]`` annotation. 
   Reference types like ``String`` and ``Date`` 
   are nullable by default, and can have the annotation without 
   failing compilation. Fields with the ``RealmObject`` type are always
   nullable and cannot be required.

Optional Fields

> ~~~~~~~~~~~~~~~

Reviewer 2: Optional fields are more or less the inverse of Required fields + have a lot of overlap in terms of content. Do you think we could combine these into one section, e.g. **Required & Optional Fields**?

In .NET, you can assign a default value in the field declaration. 

.. literalinclude:: /examples/Schemas/DefaultValues.cs
  :language: csharp

.. note:: Default Values and Nullability

   While default values ensure that a newly created object cannot contain
   a value of ``null`` (unless you specify a default value of ``null``),
   they do not impact the nullability of a field. To make a field
   non-nullable, see :ref:`Required Fields <dotnet-required-field>`.

.. dotnet-index-field:

Indexes
~~~~~~~

**Indexes** support the efficient execution of queries in
{+client-database+}. Without indexes, {+client-database+} must perform a
*collection scan*, i.e. scan every document in a collection, to select
those documents that match a query. If an appropriate index exists for a
query, {+client-database+} can use the index to limit the number of
documents that it must inspect.

Indexes are special data structures that store a small portion of a
{+realm+}'s data in an easy to traverse form. The index stores the value
of a specific field ordered by the value of the field. The ordering of
the index entries supports efficient equality matches and range-based
query operations.

> Adding an index to a field makes writing slightly slower, but makes

Reviewer 2: ```suggestion
Adding an index to a field makes write operations slightly slower, but makes
```

certain queries faster. Indexes require space in your {+realm+} file, so
adding an index to a field will increase disk space consumed by your
{+realm+} file.

You can index fields with the following types:

- ``String``
- ``ObjectId``
- ``Integer`` or ``int``
- ``Long`` or ``long``
- ``Short`` or ``short``
- ``Byte`` or ``byte[]``
- ``Boolean`` or ``bool``
- ``Date``

To index a field, use the :dotnet-sdk:`[Indexed] <reference/Realms.IndexedAttribute.html>`
annotation:

.. literalinclude:: /examples/Schemas/Indexed.cs
  :language: csharp

.. _dotnet-field-relationships:

Relationships
~~~~~~~~~~~~~

{+client-database+} supports several kinds of object relationships.

.. seealso::

   :ref:`Relationships <dotnet-client-relationships>`

.. _dotnet-field-relationships-many-to-one:

To-One

> ===========

Reviewer 2: ```suggestion
======
```

The header underline should have the same number of characters as the text

To set up a many-to-one or one-to-one relationship, create a field
with whose type is a ``RealmObject`` in your application:

.. literalinclude:: /examples/Schemas/ManyToOne.cs
  :language: csharp

Each ``Dog`` references zero or one ``Person`` instances. Nothing prevents
multiple ``Dog`` instances from referencing the same ``Person`` as an owner;
the distinction between a many-to-one and a one-to-one relationship is
up to your application.

Setting a relationship field to null removes the connection between
objects, but {+client-database+} does not delete the referenced object
unless that object is :ref:`embedded <dotnet-field-relationships-embedded>`.

.. _dotnet-field-relationships-many-to-many:

To-Many

> ============

Reviewer 2: ```suggestion
=======
```

The header underline should have the same number of characters as the text

You can create a relationship between one object and any number of objects
using a field of type ``IList<T>`` where ``T`` is a
{+service-short+} object in your application:

.. literalinclude:: /examples/Schemas/ManyToMany.cs
  :language: csharp

.. _dotnet-field-relationships-embedded:

Embedded Objects

> ================

Reviewer 2: I think we should make sure Relationships and Embedded Objects are distinct. They're both relational constructs but embedded objects are more like a complex field type and have different behavior/considerations from other relationships.

To wit, can we pull **Embedded Objects** up one level to be a peer of Relationships rather than a child?

objects of the same type in a {+realm+}. 

.. note::
   
   Once you assign a property as a primary key, you cannot change it after adding an

>    object of that type to a {+realm+}.

Reviewer: `object of that type` is redundant - once an object is added, it's primary key cannot be changed - the fact that it is of that type is not important.

   
   The following example demonstrates how to designate a primary key in an object schema:
   
   .. literalinclude:: /examples/Schemas/PrimaryKey.cs
      :language: csharp

.. _dotnet-required-optional-field:

Required and Optional Fields
~~~~~~~~~~~~~~~~~~~~~~~~~~~~

> If a field is normally nullable, you can make it required to disallow ``null`` values

Reviewer: Here and in the rest of the document field should be property - the .NET SDK doesn't work with fields.

using the ``[Required]`` annotation. In .NET, fields are nullable only if marked with with the 

> C# `? operator <https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/builtin-types/nullable-value-types>`__ except

Reviewer: `?` is not an operator.

for the following types:

- ``String``
- ``Date``
- ``ObjectId``

> - ``Decimal128``

Reviewer: This list is incorrect. It should only contain `string` and `byte[]`.

> Any value type in C# can be made nullable with the ``?``, although by

Reviewer: "the `?`" reads very awkwardly.

default they are non-nullable without it. 
Fields with the ``RealmObject`` type are always

> nullable and cannot be required.

Reviewer: "cannot be *made* required".

> You can use the ``[Required]`` annotation as seen in the following example:

Reviewer: `[Required]` is an attribute, not annotation.

 .. literalinclude:: /examples/Schemas/Required.cs
      :language: csharp

Nullable fields are optional by default in {+client-database+}, unless
otherwise specified with the :ref:`[Required] <dotnet-required-field>`

> annotation. 

Reviewer: This repeats the information in the first sentence of this section - do we really need it?

.. _dotnet-default-field-values:

Default Field Values
~~~~~~~~~~~~~~~~~~~~

You can use the built-in language features to assign a default value to a field. 

> In .NET, you can assign a default value in the field declaration. 

Reviewer: Here .NET should be C# - I know that in many places we use them interchangeably, but here we can't.

.. literalinclude:: /examples/Schemas/DefaultValues.cs
  :language: csharp

.. note:: Default Values and Nullability

   While default values ensure that a newly created object cannot contain
   a value of ``null`` (unless you specify a default value of ``null``),
   they do not impact the nullability of a field. To make a field
   non-nullable, see :ref:`Required Fields <dotnet-required-field>`.

.. dotnet-index-field:

Indexes
~~~~~~~

**Indexes** support the efficient execution of queries in
{+client-database+}. Without indexes, {+client-database+} must perform a

> *collection scan*, i.e. scan every document in a collection, to select

Reviewer: In the Realm database, "collection" is called "table" and "document" is "object". While this terminology might be more understandable for MongoDB developers, it will be confusing for people coming from Realm background.

Author: Is this something you'd liked to see changed in the .NET SDK documentation in general? It's used on other pages beside this one.

Reviewer: Gotcha - let's create a separate ticket for that then - we should probably make sure it's aligned across all SDKs.

those documents that match a query. If an appropriate index exists for a
query, {+client-database+} can use the index to limit the number of
documents that it must inspect.

Indexes are special data structures that store a small portion of a
{+realm+}'s data in an easy to traverse form. The index stores the value
of a specific field ordered by the value of the field. The ordering of
the index entries supports efficient equality matches and range-based
query operations.

Adding an index to a field makes write operations slightly slower, but makes
certain queries faster. Indexes require space in your {+realm+} file, so
adding an index to a field will increase disk space consumed by your
{+realm+} file.

You can index fields with the following types:

> - ``String``

Reviewer: `string`

> - ``String``

Reviewer: This list is missing `char`.

- ``ObjectId``
- ``Integer`` or ``int``
- ``Long`` or ``long``

> - ``Short`` or ``short``

Reviewer: Remove uppercase versions

> - ``Byte`` or ``byte[]``

Reviewer: `byte`

> - ``Boolean`` or ``bool``

Reviewer: Remove uppercase versions

> - ``Date``

Reviewer: `DateTimeOffset`

To index a field, use the :dotnet-sdk:`[Indexed] <reference/Realms.IndexedAttribute.html>`
annotation:

.. literalinclude:: /examples/Schemas/Indexed.cs
  :language: csharp

.. _dotnet-field-relationships:

Relationships
~~~~~~~~~~~~~

{+client-database+} supports several kinds of object relationships.

.. seealso::

   :ref:`Relationships <dotnet-client-relationships>`

.. _dotnet-field-relationships-many-to-one:

To-One
======

> To set up a many-to-one or one-to-one relationship, create a field

Reviewer: field -> property

> with whose type is a ``RealmObject`` in your application:

Reviewer: The type is not `RealmObject`, but rather a type that inherits from `RealmObject` or `EmbeddedObject`.

.. literalinclude:: /examples/Schemas/ManyToOne.cs
  :language: csharp

Each ``Dog`` references zero or one ``Person`` instances. Nothing prevents
multiple ``Dog`` instances from referencing the same ``Person`` as an owner;
the distinction between a many-to-one and a one-to-one relationship is
up to your application.

Setting a relationship field to null removes the connection between
objects, but {+client-database+} does not delete the referenced object
unless that object is :ref:`embedded <dotnet-field-relationships-embedded>`.

.. _dotnet-field-relationships-many-to-many:

To-Many
=======

You can create a relationship between one object and any number of objects
using a field of type ``IList<T>`` where ``T`` is a

> {+service-short+} object in your application:

Reviewer: T can be RealmObject or EmbeddedObject inheritor.

.. literalinclude:: /examples/Schemas/ManyToMany.cs
  :language: csharp

.. _dotnet-field-relationships-inverse:

Inverse Relationships
=====================

By default, {+client-database+} relationships are unidirectional. You
can follow a link from one class to a referenced class, but not in the
opposite direction. You can provide a link in the opposite direction
with the :dotnet-sdk:`[Backlink] <reference/Realms.BacklinkAttribute.html>` annotation:

.. literalinclude:: /examples/Schemas/Inverse.cs
  :language: csharp

Since relationships are many-to-one or many-to-many, following inverse
relationships can result in zero, one, or many objects.

Like any other ``RealmResults`` set, you can
:ref:`query <dotnet-client-query-engine>` an inverse relationship.

.. _dotnet-ignore-field:

Ignore a Field
~~~~~~~~~~~~~~

> If you don't want to save a field in your model to a {+realm+}, you can

Reviewer: Should mention that properties that are not autoimplemented or don't have a setter are ignored by default.

- Schema errors reported will use the internal name.

Use the :dotnet-sdk:`[MapTo] <reference/Realms.MapToAttribute.html>`
annotation to rename a field:

.. literalinclude:: /examples/Schemas/RenameField.cs
  :language: csharp

.. _dotnet-rename-object:

Rename a Class
~~~~~~~~~~~~~~

By default, {+client-database+} uses the name defined in the model class
to represent classes internally. In some cases you might want to change
this behavior:

- To support multiple model classes with the same simple name in different packages.
- To make it easier to work across platforms, since naming conventions differ.
- To use a class name that is longer than the 57 character limit enforced by {+client-database+}.
- To change a class name in .NET without forcing a migration.

Use the :dotnet-sdk:`[MapTo] <reference/Realms.MapToAttribute.html>`
annotation to rename a class:

.. literalinclude:: /examples/Schemas/RenameClass.cs
  :language: csharp

.. _dotnet-provide-subset-classes-schema:

Provide a Subset of Classes to your Realm Schema
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

By default, your application's {+service-short+} Schema includes all

> classes that extend ``RealmObject``. If you only want to include a

Reviewer: extend `RealmObject` -> inherit from `RealmObject` or `EmbeddedObject`.

> subset of classes that extend ``RealmObject`` in your {+service-short+}

Reviewer: repetition of `extend RealmObject` - can be replaced with `a subset of these classes`.
